Sources/_StringProcessing/Engine/MEBuiltins.swift

```diff
@@ -291,7 +336,7 @@ extension String {
       if isScalarSemantics {
         matched = scalar.isNewline && asciiCheck
         if matched && scalar == "\r"
-            && next != endIndex && unicodeScalars[next] == "\n" {
+            && next < end && unicodeScalars[next] == "\n" {
```

Comment:
E.g. here is a place where we implicitly rely on `end` being scalar aligned (which is a good assumption to make and would be good to assert at the very top of a match operation).